pkg/state/state_skel.go

```diff
@@ -266,12 +266,48 @@ func (s *stateSkel) handleStateObjectsDeletion(ctx context.Context) (SyncState,
 	return SyncStateIgnore, nil
 }
 
-func (s *stateSkel) deleteStateRelatedObjects(ctx context.Context) (bool, error) {
+// remove stale object of the state, returns boolean which indicates if removal is in progress and
+// an error if failed to remove an object
+func (s *stateSkel) handleStaleStateObjects(ctx context.Context,
+	currentObjs []*unstructured.Unstructured) (bool, error) {
+	reqLogger := log.FromContext(ctx)
+	reqLogger.V(consts.LogLevelInfo).Info(
+		"check state for stale objects", "State:", s.name)
+	objsToKeep := stateObjects{}
+	for _, o := range currentObjs {
+		gvkMap := objsToKeep[o.GroupVersionKind()]
+		if gvkMap == nil {
+			gvkMap = make(map[types.NamespacedName]struct{})
+		}
+		gvkMap[types.NamespacedName{Name: o.GetName(), Namespace: o.GetNamespace()}] = struct{}{}
+		objsToKeep[o.GroupVersionKind()] = gvkMap
+	}
+	found, err := s.deleteStateRelatedObjects(ctx, objsToKeep)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to delete k8s objects")
+	}
+	if found {
+		reqLogger.V(consts.LogLevelInfo).Info("removal of the state stale objects is in progress ",
+			"State:", s.name)
+		return true, nil
+	}
+	reqLogger.V(consts.LogLevelInfo).Info("no stale objects detected", "State:", s.name)
+	return false, nil
+}
+
+// is a mapping where GVK is a key and a map(set) with NamespacedNames is a value
+type stateObjects map[schema.GroupVersionKind]map[types.NamespacedName]struct{}
+
+func (s *stateSkel) deleteStateRelatedObjects(ctx context.Context, stateObjectsToKeep stateObjects) (bool, error) {
 	stateLabel := map[string]string{
 		consts.StateLabel: s.name,
 	}
 	found := false
 	for _, gvk := range getSupportedGVKs() {
+		typeObjectsToKeep := stateObjectsToKeep[gvk]
```

Comment:
while at it, consider adding a "Get" method as well.

mainly to reduce complexity.

Reply:
I added helper function for the stateObjects map